File: src/Viewport.lua
-- Skelly Viewport for tiled maps.
--
-- By Chris Herborth (https://github.com/Taffer)
-- MIT license, see LICENSE.md for details.

local Class = require 'lib/middleclass/middleclass'

local Viewport = Class('Viewport')

function Viewport:initialize(map_width, map_height, x, y, width, height)
    self.map_width = map_width
    self.map_height = map_height

    self.x = x -- These are specified in tiles, not pixels.
    self.y = y
    self.width = width or 29 -- Viewport is 29x21 in Skelly.
    self.height = height or 21
end

function Viewport:setPosition(x, y)
    if x + self.width > self.map_width then
        self.x = self.map_width - self.width
    else
        self.x = x
    end
    if y + self.height > self.map_height then
        self.y = self.map_height - self.height
    else
        self.y = y
    end
end

return Viewport

File: src/Viewport.py
# Skelly Viewport for tiled maps.
#
# By Chris Herborth (https://github.com/Taffer)
# MIT license, see LICENSE.md for details.

import pygame


class Viewport:
    ''' View into the map in tile co-ords.
    '''
    def __init__(self, map_width: int, map_height: int, rect: pygame.Rect) -> None:
        self.map_tile_width = map_width
        self.map_tile_height = map_height

        self.rect = rect

    def set_position(self, x: int, y: int) -> None:
        if x + self.rect.width > self.map_tile_width:
            self.rect.x = self.map_tile_width - self.rect.width
        else:
            self.rect.x = x

        if y + self.rect.height > self.map_tile_height:
            self.rect.y = self.map_tile_height - self.rect.height
        else:
            self.rect.y = y
